rfcs/20200505-transactional-fs.md

```diff
@@ -0,0 +1,421 @@
+
+## User Benefit
+
+With this extension proposal, users will have a more stable access to cloud storage systems as well as checkpointing redundancy can improve.
+
+## Design Proposal
+
+This RFC proposes to extend the [filesystem plugin rfc][filesystem_plugin] api with transaction markers. There can be different levels of transactions. First level can be global transactions, user starts a transaction scope. Any operations done within this scope will be grouped in this transaction. This is easiest to implement but have drawbacks such as only one transaction can be present at a time and different type of transactions may not always be reordered.
+Alternative to this approach is having multiple transaction scopes. User can create a transaction token and pass it to filesystem operations for plugin to differentiate among independent transactions. This token can be per file or per directory level granularity. Even though per file would give most flexibility, per directory level of transaction detail is most likely sufficient.
+
+Filesystem plugins may choose to ignore the transaction scopes or can delay the operations until the termination of transaction scope.
+
+### Extension to existing filesystem implementation
+
+Existing filesystem C++ api can easily be expanded by addition of three methods, an opaque structure and possibly a helper class to support transactions.
+
+```cpp
+struct TransactionToken{
+  FileSystem* owner;
+  void* id;
+};
+// C++ helper class for transaction scoping.
+template <typename T>
+class TokenScope{
+  public:
+  // Transaction name can be filename or directory name
+    TokenScope(T* fs_,const string& transaction_name):fs(fs_){
+      auto status=fs->StartTransaction(transaction_name,&token);
+    }
+    ~TokenScope(){
+      fs->EndTransaction(token);
+    }
+    TokenScope(const TokenScope&) = delete;
+    const std::unique_ptr<TransactionToken>* GetToken() const {return &token;}
+  private:
+    std::unique_ptr<TransactionToken> token;
+    T* fs;
+};
+```
+
+For a coarse granularity adding `StartTransaction` `EndTransaction` and `GetTransactionTokenForFile` methods will be sufficient. However this will prevent having multiple simultaneous transactions per file system and limit the flexibility. Thus we propose extending signature of each method with a unique pointer to `TransactionToken` structure, defaulting to `nullptr` for minimizing the impact on the existing code and allow incremental migration to implementation of transactions.
+
+```cpp
+class Filesystem {
+  // Transaction Token API extensions
+  virtual Status GetTransactionTokenForFile(const string& file_name,std::unique_ptr<TransactionToken>* token) = 0
+  virtual Status StartTransaction(const string& transaction_name, std::unique_ptr<TransactionToken>* token) = 0;
```

Comment:
std::unique_ptr<TransactionToken>* is a weird type (pointer to unique pointer to TransactionToken)

I think the signature should just be TransactionToken* and we should clearly separate methods which can create/destroy a TransactionToken from ones that just need a reference to it

Reply:
Hi @alextp Thanks for the comments. I made them as unique_ptr* to be inline with existing implementations. I don't have any special preference for it. I will change it to TransactionToken*